new credentials uploading feature

Jngai opened the pull request:
This is not done yet. There is a need for some refactoring but a major portion of it is there.

Jngai commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,7 @@
         <%= link_to new_user_credential_path, class: "btn btn-default" do %><span class="glyphicon glyphicon-plus"></span> New Credential<% end %>
         <%= link_to user_credentials_path(format: :json), class: "btn btn-default" do %><span class="glyphicon glyphicon-cloud-download"></span> Download Credentials<% end %>
+        <%= form_tag import_user_credentials_path, multipart: true, class: "btn btn-default" do %><span class="glyphicon glyphicon-upload"><%= file_field_tag :file %><%= submit_tag "Upload File"%></span><% end %>
       </div>
```
The button here looks off (words spilled over) compared to the other two buttons. Not sure if this is ok?

cantino replied:
Mind attaching a screenshot of what it looks like?

Jngai replied:
![screen shot 2016-04-11 at 1 48 06 pm](https://cloud.githubusercontent.com/assets/1547934/14436772/33ac8490-ffec-11e5-9777-82531c7ed7e3.png)

cantino replied:
I wonder if we can find a way to make this look a bit better?  Does bootstrap have a file upload button or something?

Jngai replied:
No they don't but I found a small solution and now it looks better

![screen shot 2016-04-13 at 12 05 37 pm](https://cloud.githubusercontent.com/assets/1547934/14500471/4a6505ce-0170-11e6-8e20-1131c417c6c5.png)

Jngai commented:
@cantino the spec is not ready but feel free to code review the feature.

dsander commented on app/models/user_credential.rb:
```diff
@@ -1,7 +1,7 @@
   MODES = %w[text java_script]
 
-  attr_accessible :credential_name, :credential_value, :mode
+  attr_accessible :credential_name, :credential_value, :mode, :id, :user_id, :created_at, :updated_at
```
This is dangerous, making `user_id` accessible allows to override credentials of different users.

cantino replied:
Yes, we can't add these.

dsander commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -14,6 +14,16 @@ def index
 
+  def import
+    file = params[:file]
+    UserCredential.import(file)
```
I think we should not rely on the user_id of the JSON here but enforce that the imported credentials belong to the `current_user`, so maybe `current_user.credentials.import(JSON.parse(file))`?

cantino replied:
Agreed.  Or in `UserCredential.import`, use `current_user.credentials.build(hash)`.

Jngai commented:
I did fix the issue. I also test this locally by removing `user_id` and or changing `user_id` and `removed created_at/updated_at` from uploaded file.

dsander commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -14,6 +14,19 @@ def index
+    file = params[:file]
+    content = JSON.parse(file.read)
+    content.each do |hash|
+      @user_credential = current_user.user_credentials.build(hash.except!("id")).save!
```
We have a [uniqueness constraint](https://github.com/Jngai/huginn/blob/uploadcredentialsfeature/app/models/user_credential.rb#L12) on the credential name. So we need need to handle the case where the `save` operation fails and avoid throwing an exception. You could add the imported user credentials to the association like this:
```
current_user.user_credentials << current_user.user_credentials.build(hash.except!("id"))
```
then use `if current_user.save` in the `respond_to` block and handle the failed save there.

Jngai commented:
@cantino This should be good to go.

cantino commented on spec/controllers/user_credentials_controller_spec.rb:
```diff
@@ -30,6 +31,14 @@ def valid_attributes(options = {})
     end
   end
 
+  describe "Post import" do
```
Since this feature has security implications, we should also show that credentials are not created for other users.  For example, if the `"user_id"` in the fixture is jane's id, show that credentials are *not* created for her.  Basically, show that the `"id"` and `"user_id"` in the fixture are ignored.

dsander replied:
It would also be great to add a test that ensure the redirect works when one of the credentials was not saved. When you run the test with `COVERAGE=true rake spec` you should get a test coverage report in `coverage/index.html` which shows the code paths that are not covered by tests.

cantino commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -14,6 +14,26 @@ def index
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_to user_credentials_path, notice: 'The file has already been uploaded.'}
```
I think you should show a better error message than this.  What sort of errors have you seen when saving?  Are there validation errors on `current_user`?  Or on any of the created credentials?

You could try this:

```ruby
content = JSON.parse(file.read)
new_credentials = content.map do |hash|
 current_user.user_credentials.build(hash.only(%w[credential_name credential_value mode]))
end

if new_credentials.map(&:save).all? { |save_status| save_status } # try to save all of the credentials, and check if any failed
  respond_to do |format|
    format.html { redirect_to user_credentials_path, notice: 'The file was successfully uploaded.' }
  end
else
  respond_to do |format|
    format.html { redirect_to user_credentials_path, notice: 'One or more of the uploaded credentials was not be imported due to error', status: 422 }
  end
end
```

I haven't tried this, but it seems like it would work?

Jngai commented:
@cantino ok I need another look.

Jngai commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -14,6 +14,24 @@ def index
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_to user_credentials_path, notice: 'One or more of the uploaded was not imported due to duplicate credential name error' }
```
here I removed `status: 422 ` because it was stuck on the redirect page.

Jngai commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -14,6 +14,24 @@ def index
+      end
+    else
+      respond_to do |format|
+        format.html { redirect_to user_credentials_path, notice: 'One or more of the uploaded was not imported due to duplicate credential name error', status: 422 }
```
I added , `status: 422`. I am having a redirection error locally but I am aware that this is for the server so I can't test it.

dsander commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -14,6 +14,24 @@ def index
+      current_user.user_credentials.build(hash.slice("credential_name", "credential_value", "mode"))
+    end
+
+    if new_credentials.map(&:save).all? { |save_status| save_status }
```
Is the block needed? `save` should return `true` or `false`, so `all?` should work without it.

You could also move the `if` conditional inside the `respond_to` block which would remove two lines of code :smile:

Jngai commented:
@cantino I am ready for a viewing.

Jngai commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -25,7 +25,7 @@ def import
         format.html { redirect_to user_credentials_path, notice: "The file was successfully uploaded."}
       else
-        format.html { redirect_to user_credentials_path, notice: 'One or more of the uploaded was not imported due to duplicate credential name error', status: 422 }
+        format.html { redirect_to user_credentials_path, notice: 'One or more of the uploaded was not imported due to duplicate credential name error'}
```
I decided to remove the `status: 422` symbol because according to the doc the redirection will fail. It can redirect properly only if its a `3xx error`.

Jngai commented:
coverage shows all lines of what I am doing tested.

cantino commented on spec/fixtures/user_credentials.json:
```diff
@@ -0,0 +1,20 @@
+    "id": 23,
+    "user_id": 30,
+    "credential_name": "Google_api_key",
+    "credential_value": "gcperfxrtqymqmluvskxzyiyxxfnjduzncoukyqehkrkamofwz",
```
This is fake, right? (Just checking!)

Jngai replied:
Yes its fake, I generated it using ruby map join.

cantino replied:
Nice

cantino commented on app/controllers/user_credentials_controller.rb:
```diff
@@ -14,6 +14,22 @@ def index
+      if new_credentials.map(&:save).all?
+        format.html { redirect_to user_credentials_path, notice: "The file was successfully uploaded."}
+      else
+        format.html { redirect_to user_credentials_path, notice: 'One or more of the uploaded was not imported due to duplicate credential name error'}
```
Maybe?

'One or more of the uploaded credentials was not imported due to an error.  Perhaps an existing credential had the same name?'

cantino commented:
Nice work!

Jngai commented:
@cantino I changed the notice description. Also I noticed a small bug, if a user did not choose a file to be uploaded then an error page would pop up, so I added a small rescue solution. Hope that works.

Jngai commented:
I don't think the travis-ci is me.

dsander commented:
@Jngai I think it is, compare [this](https://github.com/Jngai/huginn/blob/uploadcredentialsfeature/app/controllers/user_credentials_controller.rb#L29) with [that](https://github.com/Jngai/huginn/blob/uploadcredentialsfeature/spec/controllers/user_credentials_controller_spec.rb#L55) :smile:

Jngai commented:
@cantino I styled the file upload button please check out my newly uploaded image.

cantino commented:
I'm sorry, I'm having trouble finding the image.

Jngai commented:
I will post it agin.

Jngai commented:
@cantino here you go
![screen shot 2016-04-13 at 12 05 37 pm](https://cloud.githubusercontent.com/assets/1547934/14507535/fbe130a0-0190-11e6-83e3-a1a92909c2cc.png)

cantino commented on spec/controllers/user_credentials_controller_spec.rb:
```diff
@@ -30,6 +31,32 @@ def valid_attributes(options = {})
+      }.to change { users(:bob).user_credentials.count }.by(2)
+    end
+
+    it "asserts user credentials were created for current user only" do
```
You can probably delete this spec-- I don't think it adds anything.

Jngai commented:
@cantino deleted.

cantino commented:
I'll see if I can improve the styling a little.

dsander commented:
In my experience styling upload fields is really tricky, the only technique I know is to "hide" the input field (`visibility: hidden` not `display: none`) behind a button/div and use javascript to trigger the file selection.

Jngai commented:
Unfortunately bootstrap does not have a file upload styling.

cantino commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,7 @@
       <div class="btn-group">
         <%= link_to new_user_credential_path, class: "btn btn-default" do %><span class="glyphicon glyphicon-plus"></span> New Credential<% end %>
         <%= link_to user_credentials_path(format: :json), class: "btn btn-default" do %><span class="glyphicon glyphicon-cloud-download"></span> Download Credentials<% end %>
+        <%= form_tag import_user_credentials_path, multipart: true, class: "btn btn-default" do %><span class="glyphicon glyphicon-upload"><label class="btn btn-primary" for="file"><%= file_field_tag :file %></span><%= submit_tag "Upload File"%></label><% end %>
```
By the way, it looks like you have <span><label></span></label> which isn't valid HTML.

But, maybe we could do this?

```html
      <div class="btn-group">
        <%= link_to new_user_credential_path, class: "btn btn-default" do %><span class="glyphicon glyphicon-plus"></span> New Credential<% end %>
        <%= link_to user_credentials_path(format: :json), class: "btn btn-default" do %><span class="glyphicon glyphicon-cloud-download"></span> Download Credentials<% end %>
        <%= link_to '#', data: { toggle: 'modal', target: '#credentials-upload' }, class: "btn btn-default credentials-upload-button" do %><span class="glyphicon glyphicon-upload"></span> Upload Credentials<% end %>
      </div>

      <div id="credentials-upload" class="modal fade" tabindex="-1" role="dialog">
        <div class="modal-dialog modal-sm">
          <div class="modal-content">
            <div class="modal-header">
              <button type="button" class="close" data-dismiss="modal"><span aria-hidden="true">&times;</span><span class="sr-only">Close</span></button>
              <h4 class="modal-title">Upload Credentials</h4>
            </div>
            <div class="modal-body">
              <p>Upload a credentials file that you have previously exported from a Huginn instance.</p>
            </div>
            <div class="modal-footer">
              <%= form_tag import_user_credentials_path, multipart: true do %>
                <%= file_field_tag :file %>
                <%= button_tag 'Cancel', class: 'btn btn-default', 'data-dismiss' => 'modal' %>
                <%= submit_tag 'Upload', class: 'btn btn-primary' %>
              <% end %>
            </div>
          </div>
        </div>
      </div>
```

Jngai commented:
@cantino I just switched to the modal and now it looks like this. I know its from bootstrap. It looks about right. 
![screen shot 2016-04-19 at 10 58 01 am](https://cloud.githubusercontent.com/assets/1547934/14643814/d246476c-061d-11e6-9e3c-634436684a5d.png)

dsander commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,28 @@
+            </div>
+            <div class="modal-footer">
+              <%= form_tag import_user_credentials_path, multipart: true do %>
+                <%= file_field_tag :file %>
```
@Jngai I think that will look better when you move the `form_tag` inside of the `.modal-content` div, give the `file_field_tag` the `form-control` class and move it into `.model-body`. It should then look somehow like in the [scenario import view](https://github.com/cantino/huginn/blob/master/app/views/scenario_imports/_step_one.html.erb)

Jngai commented:
@dsander is this what you are looking for? 
![screen shot 2016-04-19 at 12 31 54 pm](https://cloud.githubusercontent.com/assets/1547934/14647222/c0a615b6-062a-11e6-9b47-f9a895f62512.png)

dsander commented:
@Jngai Yes, perfect! What do you think about using a the normal modal size by removing the `modal-sm` class?

dsander commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,28 @@
+        <div class="modal-dialog modal-sm">
+          <div class="modal-content">
+            <%= form_tag import_user_credentials_path, multipart: true do %>
+            <div class="modal-header">
```
Everything inside the `form_tag` should be indented one level more.

dsander commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,28 @@
+            <div class="modal-footer">
+              <%= button_tag 'Cancel', class: 'btn btn-default', 'data-dismiss' => 'modal' %>
+              <%= submit_tag 'Upload', class: 'btn btn-primary' %>
+            <% end %>
```
The `end` tag should be outside of the `.modal-footer` div.

Jngai commented:
@dsander I think without the `modal-sm` is better since I just noticed that the file name is actually cut off in the smaller version. I also did the minor fix that you just mentioned.

Jngai commented:
looks like this now
![screen shot 2016-04-19 at 12 59 13 pm](https://cloud.githubusercontent.com/assets/1547934/14648185/0894dcaa-062f-11e6-95e8-fb4868ddb413.png)

Jngai commented:
@cantino I just checked. I think this is good for another code review.

cantino commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,28 @@
+      <div id="credentials-upload" class="modal fade" tabindex="-1" role="dialog">
+        <div class="modal-dialog">
+          <div class="modal-content">
+              <%= form_tag import_user_credentials_path, multipart: true do %>
```
This line is indented too far.

Jngai replied:
I moved this two spaces up. I hope this is what you are looking for. I will have it nested within `<div class="modal-content">`.

Jngai replied:
ok I moved the `form_tag` 2 spaces in.

cantino commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,28 @@
+            </div>
+            <div class="modal-body">
+              <p>Upload a credentials file that you have previously exported from a Huginn instance.</p>
+                <%= file_field_tag :file, class: 'form-control' %>
```
This line is indented too far.

Jngai replied:
I moved this 2 spaces up as well.

Jngai replied:
I also moved `file_field_tag` 2 spaces in so it matches the `p`.

cantino commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,28 @@
+                <%= file_field_tag :file, class: 'form-control' %>
+            </div>
+            <div class="modal-footer">
+                <%= button_tag 'Cancel', class: 'btn btn-default', 'data-dismiss' => 'modal' %>
```
These lines are indented too far.

Jngai replied:
I just made a patch so `modal-header`, `modal-title`, `modal-body`, and `modal-footer` are all nested within the `form_tag`.

cantino commented on app/views/user_credentials/index.html.erb:
```diff
@@ -39,6 +39,28 @@
+                <%= button_tag 'Cancel', class: 'btn btn-default', 'data-dismiss' => 'modal' %>
+                <%= submit_tag 'Upload', class: 'btn btn-primary' %>
+            </div>
+              <% end %>
```
The whole block above isn't indented to match this.  What editor do you use?  Most can auto-format.

Jngai replied:
This whole block is currently within `<div class='col-md-12'>`. I am using sublime at the moment. I put them in manually.

Jngai commented:
@cantino I fix some of the indentation errors that you mentioned but I don't quite get your last comment.

Jngai commented:
@cantino I think I got it. I just put in a patch to indent/nest things correctly.

cantino commented:
Looks great!  Could you squash this?

Jngai commented:
Hey @cantino I squashed it. Thanks!

cantino commented:
Awesome, nice work @Jngai!
